File: 9-13_효율성.py
def update_board(board, r1, c1, r2, c2, degree):
    board[r1][c1] += degree
    if c2 + 1 < len(board[0]):
        board[r1][c2 + 1] -= degree
    if r2 + 1 < len(board):
        board[r2 + 1][c1] -= degree
    if r2 + 1 < len(board) and c2 + 1 < len(board[0]):
        board[r2 + 1][c2 + 1] += degree

def solution(board, skill):
    board_diff=[[0] * len(board[0]) for _ in range(len(board))]
    for type, r1, c1, r2, c2, degree in skill:
        if type == 1:
            update_board(board_diff, r1, c1, r2, c2, -degree)
        else:
            update_board(board_diff, r1, c1, r2, c2, degree)

    for r in range(len(board)):
        for c in range(1, len(board[0]), 1):
            board_diff[r][c] += board_diff[r][c - 1]

    for c in range(len(board[0])):
        for r in range(1, len(board), 1):
            board_diff[r][c] += board_diff[r - 1][c]

    answer = 0
    for r in range(len(board)):
        for c in range(len(board[0])):
            if board[r][c] + board_diff[r][c] > 0:
                answer+=1
    return answer

File: 9-13-효율성.java
import java.util.*;

class Solution {
    public static void update_board(int[][] board, int r1, int c1, int r2, int c2, int degree) {
        board[r1][c1] += degree;
        if (c2 + 1 < board[0].length)
            board[r1][c2 + 1] -= degree;
        if (r2 + 1 < board.length)
            board[r2 + 1][c1] -= degree;
        if (r2 + 1 < board.length && c2 + 1 < board[0].length)
            board[r2 + 1][c2 + 1] += degree;
    }

    public static int solution(int[][] board, int[][] skill) {
        int answer = 0;
        int[][] board_diff = new int[board.length][board[0].length];
        for (int i = 0; i < board.length; i++)
            for (int j = 0; j < board[i].length; j++)
                board_diff[i][j] = 0;
        for (int i = 0; i < skill.length; i++) {
            update_board(board_diff, skill[i][1], skill[i][2], skill[i][3], skill[i][4], skill[i][0] == 1 ? -skill[i][5] : skill[i][5]);
        }
        for (int r = 0; r < board_diff.length; r++) {
            for (int c = 1; c < board_diff[0].length; c++) {
                board_diff[r][c] += board_diff[r][c - 1];
            }
        }
        for (int c = 0; c < board_diff[0].length; c++) {
            for (int r = 1; r < board_diff.length; r++) {
                board_diff[r][c] += board_diff[r - 1][c];
            }
        }
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (board[i][j] + board_diff[i][j] > 0) answer++;
            }
        }
        return answer;
    }
}